pkg/util/kubernetes/apiserver/wpa_controller.go

```diff
@@ -38,8 +54,62 @@ func (h *AutoscalersController) RunWPA(stopCh <-chan struct{}) {
 	<-stopCh
 }
 
-// ExtendToWPAController adds the handlers to the AutoscalersController to support WPAs
-func ExtendToWPAController(h *AutoscalersController, wpaInformer v1alpha1.WatermarkPodAutoscalerInformer) {
+type checkAPI func() error
+
+func checkWPACRDAttempt(attempt *int, check checkAPI) error {
+	*attempt++
+	if err := check(); err != nil {
+		// Check if this is a known problem of missing CRD registration
+		if isWPACRDNotFoundError(err) {
+			log.Warnf("WPA CRD missing (attempt=%d): will retry", *attempt)
+			return err
+		}
+		// In all other cases return a permanent error to prevent from retrying
+		log.Errorf("WPA CRD check failed (attempt=%d): not retryable: %s", *attempt, err)
+		return backoff.Permanent(err)
+	}
+	log.Info("WPA CRD check successful")
+	return nil
+}
+
+func isWPACRDNotFoundError(err error) bool {
+	status, ok := err.(*apierrors.StatusError)
+	if !ok {
+		return false
+	}
+	reason := status.Status().Reason
+	details := status.Status().Details
+	return reason == v1.StatusReasonNotFound &&
+		details.Group == apis_v1alpha1.SchemeGroupVersion.Group &&
+		details.Kind == "watermarkpodautoscalers"
+}
+
+func checkWPACRD(wpaClient wpa_client.Interface) backoff.Operation {
+	attempt := 0
+	check := func() error {
+		_, err := wpaClient.DatadoghqV1alpha1().WatermarkPodAutoscalers(v1.NamespaceAll).List(v1.ListOptions{})
+		return err
+	}
+	return func() error {
+		return checkWPACRDAttempt(&attempt, check)
+	}
+}
+
+func waitForWPACRD(wpaClient wpa_client.Interface) {
+	exp := &backoff.ExponentialBackOff{
+		InitialInterval:     crdCheckInitialInterval,
+		RandomizationFactor: backoff.DefaultRandomizationFactor,
+		Multiplier:          crdCheckMultiplier,
+		MaxInterval:         crdCheckMaxInterval,
+		MaxElapsedTime:      0,
+		Clock:               backoff.SystemClock,
+	}
+	exp.Reset()
+	_ = backoff.Retry(checkWPACRD(wpaClient), exp)
```

Comment:
Is there a way to get the next "wait" time, so we can display it in the logs? Would be useful if people want to know if stuff is dead or just waiting for retry

Reply:
Ah yes, we could do that, though worth mentioning that we set the max of 	`crdCheckMaxInterval     = 5 * time.Minute` for this retrier.